File: Problems/Algorithms/1046. Last Stone Weight/LastStoneWeight.java
class Solution {
    public int lastStoneWeight(int[] stones) {
        int n = stones.length;
        PriorityQueue<Integer> heap = new PriorityQueue<>(Collections.reverseOrder());

        for (int stone : stones) {
            heap.offer(stone);
        }

        while (heap.size() != 1) {
            int stone1 = heap.poll();
            int stone2 = heap.poll();
            heap.offer(stone1 - stone2);
        }

        return heap.poll();
    }
}

File: Problems/Algorithms/1046. Last Stone Weight/LastStoneWeight.kt
class Solution {
    fun lastStoneWeight(stones: IntArray): Int {
        val heap = PriorityQueue<Int> {
            a, b -> b - a
        }
        stones.forEach { it -> heap.add(it) }
        
        while (heap.size > 1) {
            val first = heap.poll()
            val second = heap.poll()
            if (first > second) {
                heap.offer(first-second)
            }
        }
        
        if (heap.size == 1) {
            return heap.poll()
        }
        return 0
    }
}

